src/lib.rs

```diff
@@ -1,21 +1,392 @@
+            description("Could not create a JavaScript Runtime")
+            display("Could not create a JavaScript Runtime")
+        }
+
+        /// Could not read a value from a channel.
+        CouldNotReadValueFromChannel {
+            description("Could not read a value from a channel")
+            display("Could not read a value from a channel")
+        }
+
+        /// There was an exception in JavaScript code.
+        // TODO: stack, line, column, filename, etc
+        JavaScriptException {
+            description("JavaScript exception")
+            display("JavaScript exception")
+        }
+    }
+}
+
+impl Clone for Error {
+    fn clone(&self) -> Self {
+        self.to_string().into()
+    }
+}
 
-/// Run the main `starling` loop.
-pub fn run() -> Result<(), io::Error> {
-    let mut core = Core::new()?;
-    core.run(future::ok(()))
+/// Configuration options for building a `starling` event loop.
+///
+/// ```
+/// extern crate starling;
+///
+/// # fn foo() -> starling::Result<()> {
+/// // Construct a new `Options` builder, providing the file containing
+/// // the main JavaScript task.
+/// starling::Options::new("path/to/main.js")
+///     // Finish configuring the `Options` builder and run the event
+///     // loop!
+///     .run()?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone, Debug)]
+pub struct Options {
+    main: path::PathBuf,
+    sync_io_pool_threads: usize,
+    cpu_pool_threads: usize,
+    channel_buffer_size: usize,
 }
 
-#[test]
-fn it_works() {
+const DEFAULT_SYNC_IO_POOL_THREADS: usize = 8;
+const DEFAULT_CHANNEL_BUFFER_SIZE: usize = 4096;
+
+impl Options {
+    /// Construct a new `Options` object for configuring the `starling` event
+    /// loop.
+    ///
+    /// The given `main` JavaScript file will be evaluated as the main task.
+    pub fn new<P>(main: P) -> Options
+    where
+        P: Into<path::PathBuf>,
+    {
+        Options {
+            main: main.into(),
+            sync_io_pool_threads: DEFAULT_SYNC_IO_POOL_THREADS,
+            cpu_pool_threads: num_cpus::get(),
+            channel_buffer_size: DEFAULT_CHANNEL_BUFFER_SIZE,
+        }
+    }
+
+    /// Configure the number of threads to reserve for the synchronous IO pool.
+    ///
+    /// The synchronous IO pool is a collection of threads for shimming
+    /// synchronous IO libraries into the (otherwise completely asynchronous)
+    /// Starling system.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `threads` is 0.
+    pub fn sync_io_pool_threads(mut self, threads: usize) -> Self {
+        assert!(threads > 0);
+        self.sync_io_pool_threads = threads;
+        self
+    }
+
+    /// Configure the number of threads to reserve for the CPU pool.
+    ///
+    /// The CPU pool is a collection of worker threads for CPU-bound native Rust
+    /// tasks.
+    ///
+    /// Defaults to the number of logical CPUs on the machine.
+    ///
+    /// ### Panics
+    ///
+    /// Panics if `threads` is 0.
```

Comment:
This rubs me the wrong way... `0` doesn't make sense, so we have to deal with it *somehow*, but then saying "oh since it doesn't make sense, we should assign some other behavior to it" seems like taking a corner case and adding magic.

I know this is a very C thing to do, but I'd rather have a `reset` method or something. But really, how often do you configure a builder, and then later go back to the default configuration? Seems like the functionality isn't necessary to me...

Reply:
> But really, how often do you configure a builder, and then later go back to the default configuration?

This is a very compelling argument, so let's keep things as-is.